cpp/ql/src/Security/CWE/CWE-843/TypeConfusion.ql

```diff
@@ -0,0 +1,228 @@
+/**
+ * @name Type confusion
+ * @description Casting a value to an incompatible type can lead to undefined behavior.
+ * @kind path-problem
+ * @problem.severity warning
+ * @security-severity 9.3
+ * @precision medium
+ * @id cpp/type-confusion
+ * @tags security
+ *       external/cwe/cwe-843
+ */
+
+import cpp
+import semmle.code.cpp.dataflow.new.DataFlow
+import BadFlow::PathGraph
+
+/**
+ * Holds if `f` is a field located at byte offset `offset` in `c`.
+ *
+ * Note that predicate is recursive, so that given the following:
+ * ```cpp
+ * struct S1 {
+ *   int a;
+ *   void* b;
+ * };
+ *
+ * struct S2 {
+ *   S1 s1;
+ *   char c;
+ * };
+ * ```
+ * both `hasAFieldWithOffset(S2, s1, 0)` and `hasAFieldWithOffset(S2, a, 0)`
+ * holds.
+ */
+predicate hasAFieldWithOffset(Class c, Field f, int offset) {
+  // Let's not deal with bit fields for now.
+  not f instanceof BitField and
+  (
+    // Base case: `f` is a field in `c`.
+    f = c.getAField() and
+    offset = f.getByteOffset()
+    or
+    // Otherwise, we find the struct that is a field of `c` which then has
+    // the field `f` as a member.
+    exists(Field g |
+      g = c.getAField() and
+      // Find the field with the largest offset that's less than offset. That's
+      // the struct we need to search recursively.
+      g =
+        max(Field cand, int candOffset |
+          cand = c.getAField() and
+          candOffset = cand.getByteOffset() and
+          offset >= candOffset
+        |
+          cand order by candOffset
+        ) and
+      hasAFieldWithOffset(g.getUnspecifiedType(), f, offset - g.getByteOffset())
+    )
+  )
+}
+
+/** Holds if `f` is the last field of its declaring class. */
+predicate lastField(Field f) {
+  exists(Class c | c = f.getDeclaringType() |
+    f =
+      max(Field cand, int byteOffset |
+        cand.getDeclaringType() = c and byteOffset = f.getByteOffset()
+      |
+        cand order by byteOffset
+      )
+  )
+}
+
+/**
+ * Holds if there exists a field in `c2` at offset `offset` that's compatible
+ * with `f1`.
+ */
+bindingset[f1, offset, c2]
+pragma[inline_late]
+predicate hasCompatibleFieldAtOffset(Field f1, int offset, Class c2) {
+  exists(Field f2 | hasAFieldWithOffset(c2, f2, offset) |
+    f1.getUnspecifiedType().getSize() = f2.getUnspecifiedType().getSize()
+    or
+    lastField(f1) and
+    f1.getUnspecifiedType().getSize() <= f2.getUnspecifiedType().getSize()
+  )
+}
+
+/**
+ * Holds if `c1` is a prefix of `c2`.
+ */
+bindingset[c1, c2]
+pragma[inline_late]
+predicate prefix(Class c1, Class c2) {
+  c1.isPod() and
+  c2.isPod() and
+  if c1 instanceof Union
+  then
+    // If it's a union we just verify that one of it's variants is compatible with the other class
+    exists(Field f1, int offset | hasAFieldWithOffset(c1, f1, offset) |
+      hasCompatibleFieldAtOffset(f1, offset, c2)
+    )
```

Comment:
That's a really good point! I've added this as a FN test in 331e21bbdb5d6e51b813a11f49f585f860a39a09. We can certainly fix that, but since this will only add more results I'd prefer if we can merge this query as-is first. I'll make sure to describe this (and my proposed solution) in the issue after this is merged.

Is that okay?

Reply:
Yep that sounds like a plan, I'm keen to get some version of this merged before without too much delay.